### ISSUE TITLE
incomplete compiler flags

### ISSUE THREAD

**Issue by tru7** [linknx/linknx#8]
for years it was necessary to manually add some compiler variables for some recent compilers because the outdated configure environment was not able to derive all the necessary parameters for compiling and linking. An example is the mysql support.
I will try to solve that.

**Comment by 2franix** [comment]
Can you please detail your example?

**Comment by tru7** [comment]
sure. On my Ubuntu, compilation only works if I set this before running configure:
  export LUA_CFLAGS=`pkg-config lua5.1 --cflags`
  export LUA_LIBS=`pkg-config lua5.1 --libs`
  export LOG4CPP_LIBS=`log4cpp-config --libs`
  export LOG4CPP_CFLAGS=`log4cpp-config --cflags`
  export LIBS="-lmysqlclient_r"
all this should be automatically derived by the configure script.

**Comment by tru7** [comment]
OK, I have the code ready in my fork which improves the autoconf environment.

**Comment by 2franix** [comment]
I do not clearly remember if I had to patch my config as you mentioned to get mysql support working with v0.0.1.32. I think I had some trouble but I am unsure and don't remember the fix I used anyway...
Again, I take your conclusion for granted. Feel free to submit a PR so that I can merge with master. We'll have time to give your new config a try, then.

**Comment by 2franix** [comment]
Reopening issue as I did not mean to close it.

**Comment by tru7** [comment]
Actually it' already merged with 83d1d88dc4cec422e13e0c2baa34a828a3f67270 I was not aware that commits done in my fork after the creation of the PR also get merged at the time when the PR is executed.

**Comment by 2franix** [comment]
No worries, thanks for pointing that out.

**Comment by 2franix** [comment]
I still have an issue with the detection of the mysql lib, even after using the changes from @tru7. Here is what I did:
```
./bootstrap.sh
./configure
make check
```

The link happens to fail in the test subdirectory:
```
g++  -g -O2 -L/usr/lib/x86_64-linux-gnu -lcppunit -ldl -ldl  -o testmain testmain-ObjectControllerTest.o testmain-ObjectTest.o testmain-ObjectTest2.o testmain-TimeSpecTest.o testmain-ExceptionDaysTest.o testmain-TimerManagerTest.o testmain-PeriodicTaskTest.o testmain-XmlServerTest.o testmain-IOPortTest.o testmain-testmain.o ../src/testmain-ruleserver.o ../src/testmain-objectcontroller.o ../src/eibclient.o ../src/testmain-threads.o ../src/testmain-timermanager.o ../src/testmain-persistentstorage.o ../src/testmain-xmlserver.o ../src/testmain-smsgateway.o ../src/testmain-emailgateway.o ../src/testmain-knxconnection.o ../src/testmain-services.o ../src/testmain-suncalc.o ../src/testmain-luacondition.o ../src/testmain-ioport.o ../src/testmain-logger.o ../ticpp/libticpp.a -L/usr/local/lib -lpthsem -L/usr/lib/x86_64-linux-gnu -lcurl   -pthread -L/usr/lib/x86_64-linux-gnu -lesmtp -lssl -lcrypto -ldl 
../src/testmain-persistentstorage.o: dans la fonction « MysqlPersistentStorage::~MysqlPersistentStorage() »:
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:171: référence indéfinie vers « mysql_close »
../src/testmain-persistentstorage.o: dans la fonction « MysqlPersistentStorage::MysqlPersistentStorage(ticpp::Element*) »:
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:145: référence indéfinie vers « mysql_init »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:149: référence indéfinie vers « mysql_options »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:156: référence indéfinie vers « mysql_real_connect »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:153: référence indéfinie vers « mysql_options »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:165: référence indéfinie vers « mysql_options »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:159: référence indéfinie vers « mysql_error »
../src/testmain-persistentstorage.o: dans la fonction « MysqlPersistentStorage::writelog(std::string const&, std::string const&) »:
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:255: référence indéfinie vers « mysql_real_query »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:257: référence indéfinie vers « mysql_error »
../src/testmain-persistentstorage.o: dans la fonction « MysqlPersistentStorage::write(std::string const&, std::string const&) »:
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:196: référence indéfinie vers « mysql_real_query »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:198: référence indéfinie vers « mysql_error »
../src/testmain-persistentstorage.o: dans la fonction « MysqlPersistentStorage::read(std::string const&, std::string const&) »:
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:215: référence indéfinie vers « mysql_real_query »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:220: référence indéfinie vers « mysql_store_result »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:222: référence indéfinie vers « mysql_num_rows »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:232: référence indéfinie vers « mysql_free_result »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:238: référence indéfinie vers « mysql_error »
/home/cyrille/documents/develop/linknx/test/../src/persistentstorage.cpp:224: référence indéfinie vers « mysql_fetch_row »
```

I tried to run the same command line with an extra `-lmysqlclient` and it succeeded.

Do you have any insight as to why autoconf is not able to detect this dependency?

**Comment by tru7** [comment]
configure fully relies on mysql_config which is typically part of libmysqlclient-dev to find and set library flags.
what is the output of "mysql_config --cflags" and "mysql_config --libs"?
what ist the output when checking for mysql?
how do you enable mysql in the configure command?
do you have private settings for CFLAGS and LIBS?

Strange thing is: when mysql cannot be found it should be disabled completely. Maybe this part is not perfect yet.

**Comment by tru7** [comment]
@2franix I had another look at your error report and I realized that I never have compiled the code in the test subdir because I didn't have cppunit installed and I didn't know of "make check". Now after installing libcppunit-dev and running "make check" I can see the same problem. Some more stuff has to be fixed, also some source code to get it compiled successfully. But I think I got it already.
I have created a branch named fix-testmain in my fork at https://github.com/tru7/linknx.git. Can you test this please?

**Comment by 2franix** [comment]
mysql_config --cflags and --libs look good:
-I/usr/include/mysql -DBIG_JOINS=1  -fno-strict-aliasing    -g -DNDEBUG
-L/usr/lib/x86_64-linux-gnu -lmysqlclient -lpthread -lz -lm -ldl

When configure checks for mysql (excerpt from config.log):
configure:6377: checking for mysql_config
configure:6393: found /usr/bin/mysql_config
configure:6405: result: mysql_config
configure:6423: checking for mysql
configure:6425: result: yes

mysql is enabled by means of configure --with-mysql. But I also tried 
invoking configure without parameters, which led to the same behaviour.

Regarding your question about private CFLAGS / LIBS, I am not sure in 
which file I should look in. But my build system is pristine: I ran 
bootstrap.sh and did not touch anything that was generated.

I noticed that the "-lmysqlclient" is passed when linking the linknx 
executable in the src subfolder. But it is not when linking the testmain 
executable in the test subfolder. My feeling is this is a 
misconfiguration of automake or something like that. But I am lacking 
knowledge about this tool to figure it out. I am in the process of 
reading the GNU autotools manual but that will take a few months before 
I get familiar with this build system...

Would you have time to try to reproduce this build issue on your end? 
Try to do the same as me: get a fresh copy of the source, run 
bootstrap.sh, then "make check" in the root folder. Let me know what 
happens.

Thanks a lot for your assistance, anyway. Greatly appreciated.

**Comment by 2franix** [comment]
Ok, forget my last comment. I had not seen your last comment when I submitted it. For some reason, I do not always get a notification by email. Since I am relying on that to catch new activity on this project, I wasted my time tonight.

Anyway, your last comment is somehow a confirmation of what I suspected. Thanks a lot for looking into it, @tru7.
I will give your fix a try ASAP, for sure! I'll keep you posted.

Thanks again. This GNU autotools is really giving me fits!

BTW, I reopen the issue until this is completely fixed.

**Comment by tru7** [comment]
@2franix I'm kind of new to autoconf myself but it's pretty easy to understand: all you have to deal with are the Makefile.am and configure.ac files. All other stuff is generated automatically. It's quite easy to read existing code, it's a little more difficult to fix code (I spent many hours) and of course it's rather difficult to write this stuff from scratch. But there are many examples.
All I bascially did so far is fixing some Makefile.am files and the configure.ac, but also a few source code files to get around compiler errors.
In my first attempts I never realized the part in the test subdir. For all those years I neve compiled the testmain part. But after your error report I dealt with it yesterday and it looks good to me in my testing branch. I needed to fix the Makefile.am to handle the -lmysqlclient and the -lcppunit correctly.

Once we are done with all the autoconf updates we should increase version to 0.0.1.33. I can take care of that in the next pull request and add some info to changelog and authors.

**Comment by 2franix** [comment]
My understanding of the autotools is quite the same as yours. And as long as I do not have strong knowledge about it, I prefer to act very cautiously, not to break anything.
I still have to merge your fix. This is going to be done as soon as real life gives me some time...

Regarding the version bump, I keep it in mind but I think there is no urgency. The version has to be increased before we setup a new release. For now, master is a development branch that no one is expected to deploy for production. My plan is to integrate my own fixes (see the open issues) and consolidate tests to lower the risk of regression (only Jef was an expert, so far). Then, we can consider releasing v1. After so many years, I think that it is acceptable to get rid of this v0...

As soon as tests can run again, I will only accept pull requests that do not break tests and that include testing of the own functionality added by the PR.

For the version, this has to be fixed in configure.ac (cf. AC_INIT(linknx, 0.0.1.32, linknx@ouaye.net)), right?

**Comment by tru7** [comment]
@2franix Yes, version setting is done in the AC_INIT macro.
I would like to have a separate version 0.0.1.33 with all my fixes for autoconf and some source files. It's basically version 0.0.1.32 specifically fixed to get a clean compilation like the -lmysqlclient part, but without any functional change.
This is important for me because I want to use the githup repo instead of sourceforge for building the package for OpenWrt including my fixes (once I'm finished with it). This has to be a higer version to build a package different from the current version. Currently a few extra tweaks are necessary to get 0.0.1.32 compiled for OpenWrt.
So I would like to have a distinct version to refer to based on my fixes before you start changing source code. Is that OK with you?

**Comment by 2franix** [comment]
Yes, that makes sense. If you need an intermediary version, we'll do it.

**Comment by tru7** [comment]
@energy01 Fine. I'm just working on it. I will redo the fix-testmain part. This will take care of the -lmysqlclient in test subdir and include support for cppunit over the whole files. I will change version and include info in AUTHORS.

**Comment by tru7** [comment]
Uuups @2franix sent to the wrong one: Fine. I'm just working on it. I will redo the fix-testmain part. This will take care of the -lmysqlclient in test subdir and include support for cppunit over the whole files. I will change version and include info in AUTHORS.

**Comment by 2franix** [comment]
Fixed by PR #13.
